### PR TITLE
Enhancement dao CUD

### PR DESCRIPTION
**Pull request by kiserd**
Added several CRUD functions to the dao file. I'm certain we will come across improvements as we integrate them into the application. We can create issues as we go.

### CODE REVIEW

File: portfolio-project-matching-app/Firebase/clientApp.ts [hunk 1]
```diff
@@ -1,7 +1,25 @@
 // Import the functions you need from the SDKs you need
 import { initializeApp } from "firebase/app";
 // TODO: Add SDKs for Firebase products that you want to use
-import { getFirestore, collection, query, where, doc, getDoc, getDocs, addDoc, setDoc } from 'firebase/firestore';
+import {
+  // REFERENCE GETTERS
```
Comment (kiserd) [inline portfolio-project-matching-app/Firebase/clientApp.ts:5]: Did my best to organize my imports and exports. There are a ton and things can easily get messy.

File: portfolio-project-matching-app/Firebase/clientApp.ts [hunk 2]
```diff
@@ -19,22 +37,63 @@ const firebaseApp = initializeApp(firebaseConfig);
...
 
-  RETURN:       snapshot of document with provided document ID
+                  addNewDoc('technologies', {name: _, language: _, ...})
+
+    RETURN:       reference to new document
   */
-  const docRef = doc(db, collectionName, docId);
-  const querySnapshot = await getDoc(docRef);
-  return querySnapshot;
+  const collRef = collection(db, coll);
```
Comment (kiserd) [inline portfolio-project-matching-app/Firebase/clientApp.ts:58]: Tried my best to go through and make my variable names a little more consistent. I could still do a ton of work in this regard. I was really learning Firebase as I went and didn't necessarily understand the nuance of references / snapshots / etc.

File: portfolio-project-matching-app/backend/dao.js [hunk 3]
```diff
@@ -1,246 +1,351 @@
...
+} from '../backend/daoUser'
+import { DocumentReference } from '@firebase/firestore'
 
 /*
-    PROJECT LEVEL QUERIES
+    CREATE
 */
 
-const getAllProjects = async () => {
+const createDoc = async (coll, payload) => {
```
Comment (kiserd) [inline portfolio-project-matching-app/backend/dao.js:47]: This was one of my newer consolidated / modularized queries that got rid of several `createProject` / `createUser` type queries.

File: portfolio-project-matching-app/backend/dao.js [hunk 4]
```diff
@@ -1,246 +1,351 @@
...
-    return users;
 }
 
-/*
-    USER LEVEL QUERIES
-*/
+// helper functions, don't export
 
-const getAllUsers = async () => {
+const getPayload = (coll, id1, id2) => {
```
Comment (kiserd) [inline portfolio-project-matching-app/backend/dao.js:160]: Little helper function to keep things cleaner in the main function. I could do a ton more work leveraging helpers in tidying things up a bit.

File: portfolio-project-matching-app/backend/dao.js [hunk 5]
```diff
@@ -1,246 +1,351 @@
...
+// see daoProject.js, daoUser.js, and daoTechnology.js
 
-    return user;
-}
+/*
+    UPDATE
+*/
 
-const getProjectsByUserId = async (userId) => {
+const updateDoc = async (coll, id, payload) => {
```
Comment (kiserd) [inline portfolio-project-matching-app/backend/dao.js:193]: Same thing here - got rid of several dedicated `update` functions.

File: portfolio-project-matching-app/backend/daoProject.js [hunk 6]
```diff
@@ -0,0 +1,482 @@
...
+    // UPDATE
+    updateDocument,
+    // DELETE
+    deleteDocById,
+} from '../Firebase/clientApp.ts'
+import { User } from '../models/User'
+import { Project } from '../models/Project'
+import { Technology } from '../models/Technology'
+
+/*
```
Comment (kiserd) [inline portfolio-project-matching-app/backend/daoProject.js:19]: The majority of these functions will remain unused. I've imported many of the `READ` type functions into `dao.js` along with the `createLike` and `deleteLike` functions. Other than that, they probably won't be used, but we might not want to delete them quite yet - they could come in handy.

File: portfolio-project-matching-app/backend/daoTechnology.js [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    getCollectionSnapshot,
+    getDocSnapshotById,
+} from '../Firebase/clientApp.ts'
+import { Technology } from '../models/Technology'
+
+/*
+    READ
+*/
+
+const getAllTechnologies = async () => {
```
Comment (kiserd) [inline portfolio-project-matching-app/backend/daoTechnology.js:11]: These two `READ` type functions are ones that we will most likely want to use, I haven't quite thought of a clever way to modularize these functions without simply passing the document snapshot straight to the calling function (which could potentially work, but would lose the association arrays).

File: portfolio-project-matching-app/backend/daoUser.js [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+} from '../Firebase/clientApp.ts'
+import { User } from '../models/User'
+import { Project } from '../models/Project'
+import { Technology } from '../models/Technology'
+
+/*
+    CREATE
+*/
+
+const createNewUserDoc = async (user) => {
```
Comment (kiserd) [inline portfolio-project-matching-app/backend/daoUser.js:21]: Once again, a good hunk of these are redundant after consolidating in `dao.js`.

File: portfolio-project-matching-app/models/Project.js [hunk 9]
```diff
@@ -1,20 +1,88 @@
...
-        this.capacity = docSnapshot.data().capacity;
-        this.census = docSnapshot.data().census;
-        this.open = docSnapshot.data().open;
-        this.likes = docSnapshot.data().likes;
-        this.ownerId = docSnapshot.data().owner.id;
-        this.owner = null;
-        this.technologies = null;
-        this.users = null;
+    // static factory methods (used like overloaded constructors)
+    static fromDocSnapshot(id, docSnapshot) {
```
Comment (kiserd) [inline portfolio-project-matching-app/models/Project.js:16]: I added some static factory methods in place of having a single constructor. This gives the developer a bit more freedom when creating one of our custom objects / classes.

File: portfolio-project-matching-app/models/Technology.js [hunk 10]
```diff
@@ -1,15 +1,54 @@
+class Technology {
+    // fields
+    id;
+    name;
+    language;
+    backend;
 
+    // static factory methods (used like overloaded constructors)
+    static fromDocSnapshot(id, docSnapshot) {
```
Comment (kiserd) [inline portfolio-project-matching-app/models/Technology.js:9]: I added some static factory methods in place of having a single constructor. This gives the developer a bit more freedom when creating one of our custom objects / classes.

File: portfolio-project-matching-app/models/User.js [hunk 11]
```diff
@@ -1,15 +1,85 @@
...
 
-    constructor(docSnapshot) {
-        this.id = docSnapshot.id;
-        this.email = docSnapshot.data().email;
-        this.username = docSnapshot.data().username;
-        this.introduction = docSnapshot.data().introduction;
-        this.projects = null;
-        this.technologies = null;
+    // static factory methods (used like overloaded constructors)
+    static fromDocSnapshot(id, docSnapshot) {
```
Comment (kiserd) [inline portfolio-project-matching-app/models/User.js:11]: I added some static factory methods in place of having a single constructor. This gives the developer a bit more freedom when creating one of our custom objects / classes.

File: portfolio-project-matching-app/pages/test_dao.js [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+    deleteProjectsUsersDoc,
+    deleteUserDoc,
+    getProjectById,
+    updateDoc,
+    updateProject,
+} from '../backend/dao'
+import { Project } from '../models/Project'
+import { User } from '../models/User'
+
+const test_createProject = () => {
```
Comment (kiserd) [inline portfolio-project-matching-app/pages/test_dao.js:22]: We will certainly end up deleting this. I ran some very manual and very basic tests on the work I was doing.

**Review by Josh-Kaiser-OSU — APPROVED** [review]

**Review by ylijokic — APPROVED** [review]

**Pull request merged**